core/src/main/scala/laserdisc/protocol/RESP.scala

```diff
@@ -393,48 +393,75 @@ sealed trait RESPFunctions { this: RESPCodecs =>
 
   final def stateOf: BitVector => String | BitVectorState =
     bits => bits.consume(BitsInByte) {
-      case `plus`
-           | `minus`
-           | `colon` => Right(No)
-      case `dollar`  => Right(Bulk)
-      case `star`    => Right(Multi)
-      case other     => Left(s"unidentified RESP type (Hex: ${other.toHex})")
+      case `plus`   => Right(SimpleString)
+      case `minus`  => Right(Error)
+      case `colon`  => Right(Integer)
+      case `dollar` => Right(BulkString)
+      case `star`   => Right(RedisArray)
+      case other    => Left(s"unidentified RESP type (Hex: ${other.toHex})")
     }
     .flatMap {
-      case (remainder, Bulk) =>
+      case (remainder, BulkString) =>
         evalWithSizeDecodedFrom(remainder) {
-          ds =>
-            if (ds.value >= 0 && ds.remainder.size == (crlf.size + (ds.value * BitsInByte))) CompleteVector // Complete: expected size
-            else if (ds.value == -1) CompleteVector // Complete: empty bulk
-            else MissingBits((crlf.size + (ds.value * BitsInByte)) - ds.remainder.size)
+          case Left(_)   => IncompleteVector
+          case Right(ds) =>
+            val expectedBulkSize = (ds.value * BitsInByte) + crlf.size
+
+            if (ds.value >= 0 && ds.remainder.size == expectedBulkSize)
+              CompleteVector // Complete: expected size
+
+            else if (ds.value >= 0 && ds.remainder.size > expectedBulkSize) {
+              val completeBulkSize = bits.bytes.indexOfSlice(crlfBytes, 0L) * BitsInByte + crlf.size + expectedBulkSize
+              CompleteWithRemainder(bits.take(completeBulkSize), bits.drop(completeBulkSize)) // Complete plus some remainder
+            }
+
+            else if (ds.value == -1 && ds.remainder.size == 0)
+              CompleteVector // Complete: empty bulk
+
+            else if (ds.value == -1 && ds.remainder.size > 0) {
+              val completeBulkSize = 3 * BitsInByte + crlf.size
+              CompleteWithRemainder(bits.take(completeBulkSize), bits.drop(completeBulkSize)) // Complete plus some remainder
+            }
+
+            else
+              MissingBits(expectedBulkSize - ds.remainder.size)
         }
 
-      case (remainder, Multi) =>
+      case (remainder, RedisArray) =>
         arrayCodec.decode(remainder).fold(
           _ => Right(IncompleteVector),
           r => Right(CompleteAndDecoded(r))
         )
 
-      case (_, No)    => Right(CompleteVector)
+      case (_, SimpleString)
```

Comment:
do we need to distinguish the 3?
also, what about `case (_, SimpleString | Integer | Error) => Right(CompleteVector)`?

Reply:
I will refactor it. I will verify in another PR if the three of them are affected by the same issue I'm fixing here. If they are not needed I will remove them then.